README.md
[![Contributors][contributors-shield]][contributors-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![DOI][zenodo-shield]][zenodo-url]
[![CodeFactor][codefactor-shield]][codefactor-url]
[![Maintainability Rating][maintainability-shield]][sonarcloud-url]
[![Security Rating][security-shield]][sonarcloud-url]
[![Vulnerabilities][vulnerabilities-shield]][sonarcloud-url]
[![Technical Debt][technical-debt-shield]][sonarcloud-url]
[![Storybook][storybook-shield]][storybook-url]
[![Curated with FAIRshare][fairshare-shield]][fairshare-url]

<!-- HEADER -->
<br />
<p align="center">
  <a href="#">
    <img src="/src/assets/brand/logo-with-name.png" alt="Logo" width="400">
  </a>

  <h3 align="center">FAIR data and software sharing made easy</h3>

  <p align="center">
    Your one-stop tool for rapidly curating and sharing biomedical research data and software according to applicable FAIR guidelines <br/>
    <br />
    <br />
    <a href="https://docs.fairshareapp.io/docs/intro"><strong>Explore our documentation »</strong></a>
    <br />
    <br />
    <a href="https://github.com/fairdataihub/FAIRshare/issues/new/choose">Report Issue</a>
    ·
    <a href="https://fairdataihub.org/contact-us">Submit feedback </a>
  </p>
</p>

<!-- TABLE OF CONTENTS -->

## Table of Contents

- [About](#about)
- [Downloading FAIRshare](#Downloading-FAIRshare)
- [Documentation](#Documentation)
- [Issues and Feedback](#Issues-and-Feedback)
- [Using the Source Code](#Using-the-Source-Code)
- [License](#license)
- [Privacy Policy](#Privacy-Policy)
- [Acknowledgements](#acknowledgements)

## About

FAIRshare is a cross-platform desktop software that allows researchers to easily organize and share their biomedical research data and software according to applicable FAIR guidelines. In our first phase of development, we are implementing a process for making biomedical research software FAIR and also working on supporting the FAIRification of COVID-19 and other infectious disease related research data (immunology, genomics, epidemiology, etc.). Learn more about FAIRshare, its impact, the development timeline, and related resources on our group's website at https://fairdataihub.org/fairshare.

<p align="center">
  <img src="https://github.com/fairdataihub/fairdataihub-website/blob/main/public/images/hero/fairshare-macos.png" alt="interface" width="600">
  <br/>
  <i> A screenshot of the user interface of FAIRshare. </i>
  </img>
</p>

## Downloading FAIRshare

FAIRshare is distributed as an easy-to-install desktop application for Windows, Mac OS, and Ubuntu. Follow the instructions provided in [our dedicated documentation](https://docs.fairshareapp.io/docs/getting-started/download-fairshare) to download the version suitable for your operating system.

All our releases are also accessible in the [releases tab](https://github.com/fairdataihub/FAIRshare/releases/latest).

## Documentation

See the [FAIRshare documentation](https://docs.fairshareapp.io/docs/intro) for details about each features.

## Issues and Feedback

To report any issues with the software, suggest improvements, or request a new feature, please open a new issue via the [Issues tab](https://github.com/fairdataihub/FAIRshare/issues). Provide adequate information (operating system, steps leading to error, screenshots) so we can help you efficiently. Alternatively, you could also use [our feedback form](https://fairdataihub.org/contact-us).

## Using the Source Code

If you want to use the source code, contribute, modify FAIRshare to better suit your needs, or compile it for an operating system not mentioned above, please follow the instructions provided in the [developer guide](https://docs.fairshareapp.io/docs/developer-documentation/project-setup) of our documentation.

## License

FAIRshare is distributed under the MIT License. See [LICENSE](https://github.com/fairdataihub/FAIRshare/blob/main/LICENSE) for more information.

## Privacy Policy

FAIRshare uses Google Analytics to track user interactions with the application. Tracking is anonymous and no identifying personal information is collected. Tracking is implemented as part to report usage statistics to the funders of FAIRshare. Tracking is also used to identify errors and subsequently improve user experience. To learn more about what is tracked and how we use Google Analytics within FAIRshare, please read our [Privacy Policy](https://docs.fairshareapp.io/docs/privacypolicy).

## Acknowledgements

- [NIAID](https://www.niaid.nih.gov/) (Funding source)
- [The Butte Lab](https://buttelab.ucsf.edu/) (Collaborators)

[contributors-shield]: https://img.shields.io/github/contributors/fairdataihub/FAIRshare.svg?style=flat-square
[contributors-url]: https://github.com/fairdataihub/FAIRshare/graphs/contributors
[stars-shield]: https://img.shields.io/github/stars/fairdataihub/FAIRshare.svg?style=flat-square
[stars-url]: https://github.com/fairdataihub/FAIRshare/stargazers
[issues-shield]: https://img.shields.io/github/issues/fairdataihub/FAIRshare.svg?style=flat-square
[issues-url]: https://github.com/fairdataihub/FAIRshare/issues
[license-shield]: https://img.shields.io/github/license/fairdataihub/FAIRshare.svg?style=flat-square
[license-url]: https://github.com/fairdataihub/FAIRshare/blob/main/LICENSE
[zenodo-shield]: https://zenodo.org/badge/DOI/10.5281/zenodo.6369429.svg
[zenodo-url]: https://doi.org/10.5281/zenodo.6369429
[lgtm-alerts-shield]: https://img.shields.io/lgtm/alerts/g/fairdataihub/FAIRshare.svg?logo=lgtm&logoWidth=18
[lgtm-alerts-url]: https://lgtm.com/projects/g/fairdataihub/FAIRshare/alerts/
[lgtm-javascript-shield]: https://img.shields.io/lgtm/grade/javascript/g/fairdataihub/FAIRshare.svg?logo=lgtm&logoWidth=18
[lgtm-javascript-url]: https://lgtm.com/projects/g/fairdataihub/FAIRshare/context:javascript
[lgtm-python-shield]: https://img.shields.io/lgtm/grade/python/g/fairdataihub/FAIRshare.svg?logo=lgtm&logoWidth=18
[lgtm-python-url]: https://lgtm.com/projects/g/fairdataihub/FAIRshare/context:python
[codefactor-shield]: https://www.codefactor.io/repository/github/fairdataihub/fairshare/badge
[codefactor-url]: https://www.codefactor.io/repository/github/fairdataihub/fairshare
[technical-debt-shield]: https://sonarcloud.io/api/project_badges/measure?project=fairdataihub_FAIRshare&metric=sqale_index
[maintainability-shield]: https://sonarcloud.io/api/project_badges/measure?project=fairdataihub_FAIRshare&metric=sqale_rating
[security-shield]: https://sonarcloud.io/api/project_badges/measure?project=fairdataihub_FAIRshare&metric=security_rating
[vulnerabilities-shield]: https://sonarcloud.io/api/project_badges/measure?project=fairdataihub_FAIRshare&metric=vulnerabilities
[sonarcloud-url]: https://sonarcloud.io/summary/new_code?id=fairdataihub_FAIRshare
[storybook-shield]: https://raw.githubusercontent.com/storybooks/brand/master/badge/badge-storybook.svg
[storybook-url]: https://main--628e928cd2515a004ad2f0b7.chromatic.com/
[fairshare-shield]: https://raw.githubusercontent.com/fairdataihub/FAIRshare/main/badge.svg
[fairshare-url]: https://fairdataihub.org/fairshare


CHANGELOG.md
# Changelog

All notable changes to `FAIRshare` will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## v2.1.0 - 2023-07-04

### Feature additions

- Added new workflows for both NCBI GEO and Immport. These repositories are not enabled for users but will be in the future.
- Added support for better user validation on Figshare.

### Bug fixes

- The bio.tools register button is now only shown if the data type is an allowed application.
- The text in the `Review standards` section has been revised to better reflect the purpose of the project.
- Electron has been updated to v25.

## v2.0.0 - 2022-08-30

### Feature additions

- The GitHub release workflow has been reworked to not use webhooks from the repository. You can now control what you post to your data repository with this new workflow. You are also able to add additional files directly from your system or a GitHub release to your final deposition.
- A new authentication system has been developed for FAIRshare. At the moment, the GitHub login flow is the only supported authentication method. This will be expanded in the future to support other accounts such as Zenodo and Figshare.
- When you reorder items in draggable lists(such as `authors` and `keywords`), a new animation is now shown to signify the change and to make it easier to see where you are moving the item.
- Added an option for the user to login to Figshare from directly within their workflow.
- Added a dataset preview option for the figshare workflow.

### Bug fixes

- Removed a button that was accidentally added to the `GitHub` publish page.
- Fixed an issue where the `Contributors` section was not collapsed when a new Zenodo dataset was loaded.
- Replaces the server presence check function timing from exponential backoff to a fixed interval. Now the front end portion of FAIRshare will check at startup every second for 30 seconds to see if the server is up. If the server is not up after 30 seconds, the front end will display an error message.
- Reverted Headless UI to v1.6 for the message box transitions to work properly.
- Fixed an issue where the correct upload type was not being assigned for the Zenodo workflow.
- Fix code smells reported by SonarCloud.

## v1.4.0 - 2022-06-03

### Feature additions

- You can now submit your datasets to Figshare using FAIRshare. This is currently only enabled for local and new datasets. Support for GitHub repos and for publishing new versions of datasets to Figshare is coming soon.
- Added support for the ability to select `Upload type` for Zenodo.
- Storybook integration has been added for components used within FAIRshare. View the [library](https://www.chromatic.com/library?appId=628e928cd2515a004ad2f0b7) and the [storybook](https://628e928cd2515a004ad2f0b7-jmdpzjjikc.chromatic.com/).

### Bug fixes

- FAIRshare will now focus on the currently open app instance if a user opens multiple instances of FAIRshare. This should prevent issues that can arise from the server side.
- The `Curated with FAIRshare` badge has a new url that we can support better.
- Updated Zenodo metadata text to better fit the context of the form.

## v1.3.1 - 2022-05-18

### Bug Fixes

- Fixed an issue where the codemeta.json had an incorrect key for the first release date.

## v1.3.0 - 2022-04-15

### Feature additions

- Add support for a user to submit their own unique identifier in the `codemeta` for a dataset.
- Switch the server check full screen loading spinner to a card style notification for better usability.
- Added a way to notify the user if something went wrong with the update process for the backend.
- You can now register your software application on [bio.tools](https://bio.tools) directly using FAIRshare. This is a great way to have your application in a public registry.
- FAIRshare now has badges! Wohoo! When you finish submitting your data you will be prompted with a badge that you can add to your GitHub repository.

### Bug fixes

- Fixed an issue where a tool to show the size of the screen was being displayed on the app if the backend server didn't load fast enough.
- Fixed an issue where buttons provided by Element Plus were being overwritten by Tailwind's css reset.
- Fixed an issue where a comment added by FAIRshare to the `CITATION.cff` file was not being shown in the UI.
- Fixed some bugs where the wrong keyword was being displayed when previewing metadata files.
- Updated some analytics events to better understand who is using our software for future grant applications.

## v1.2.0 - 2022-03-31

### Feature additions

- You will now be warned when you select a GitHub repo that doesn't have the appropriate permissions. This should prevent you from accidentally uploading to a repo that you don't have permission to.
- Added a button to let you view the repo if you decide to not publish with FAIRshare.
- Added support for reading pre-existing `codemeta` files directly from your GitHub repository.

### Bug fixes

- Fixed an issue where app log files were not being created in the correct location.
- Fixed an issue where incorrect file keys were being read from the `codemeta.json` file.
- Fixed some issues where certain GitHub repositories were not returning a license correctly. This was causing FAIRshare to assume an invalid license existed on the repository.
- Fixed an issue where an publishing a new version of a Zenodo deposition where no orcids for authors were provided could cause the app to freeze.

## v1.1.0 - 2022-03-21

### Feature additions:

- Added indicators to show when FAIRshare is updating to a new version.

### Bug fixes:

- On Linux the backend server was not being bundled correctly. This has been fixed to start on app launch.
- Fixed an issue where the splash screen on macOS was rendering a shadow.

## v1.0.0 - 2022-03-18

### Feature additions:


src/background.js
"use strict";

import { app, protocol, BrowserWindow, ipcMain, Menu, shell, dialog } from "electron";
import { enable as enableWebContents } from "@electron/remote/main";
import { createProtocol } from "vue-cli-plugin-electron-builder/lib";
import { autoUpdater } from "electron-updater";
import installExtension, { VUEJS3_DEVTOOLS } from "electron-devtools-installer";

import axios from "axios";
import uniqid from "uniqid";

import menuTemplate from "./scripts/menu";
import "./scripts/utilities";
import "./scripts/auth";
import "./scripts/dialogs";

const fp = require("find-free-port");
const fs = require("fs-extra");
const path = require("path");
const log = require("electron-log");
const nodeUrl = require("url");

const USER_PATH = app.getPath("home");
const CONFIG_STORE_PATH = path.join(USER_PATH, ".fairshare", "config.json");

const SESSION_ID = uniqid();

log.transports.file.resolvePath = () => path.join(USER_PATH, ".fairshare", "logs", "app.log");
log.transports.console.format = "{h}:{i}:{s} {text}";

log.info("starting log");

if (process.platform === "win32") {
  app.setAppUserModelId(app.name);
}

const getReleaseChannel = () => {
  const exists = fs.pathExistsSync(CONFIG_STORE_PATH);

  if (!exists) {
    return "latest";
  } else {
    try {
      let config = fs.readJsonSync(CONFIG_STORE_PATH);
      if ("releaseChannel" in config) {
        if (config.releaseChannel == "beta") {
          return "beta";
        } else {
          return "latest";
        }
      }
    } catch (err) {
      console.error(err);
      return "latest";
    }
  }
};

const updateChannel = getReleaseChannel();

autoUpdater.channel = updateChannel;
autoUpdater.logger = require("electron-log");
autoUpdater.logger.transports.file.level = "info";

const isDevelopment = process.env.NODE_ENV !== "production";

require("@electron/remote/main").initialize();

// Scheme must be registered before the app is ready
protocol.registerSchemesAsPrivileged([
  { scheme: "app", privileges: { secure: true, standard: true } },
]);

const PY_FOLDER = "pyflask";
const PY_MODULE = "api";
let mainWindow;

let pyProc = null;
let pyPort = null;

global.PY_PORT = pyPort;

const startingPort = 7632;
const portRange = 100;

const guessPackaged = () => {
  const unixPath = path.join(process.resourcesPath, PY_MODULE);
  const winPath = path.join(process.resourcesPath, PY_MODULE + ".exe");

  return require("fs").existsSync(unixPath) || require("fs").existsSync(winPath);
};

// check if the python dist folder exists
const getScriptPath = () => {
  if (!guessPackaged()) {
    return path.join(__dirname, PY_FOLDER, PY_MODULE + ".py");
  }

  if (process.platform === "win32") {
    return path.join(process.resourcesPath, PY_MODULE + ".exe");
  }

  return path.join(process.resourcesPath, PY_MODULE);
};

const killAllPreviousProcesses = async () => {
  console.log("Killing all previous processes");

  // kill all previous python processes that could be running.
  let promisesArray = [];

  // create a loop of 100
  for (let i = 0; i < portRange; i++) {
    promisesArray.push(
      axios.post(`http://127.0.0.1:${startingPort + i}/fairshare_server_shutdown`, {})
    );
  }

  // wait for all the promises to resolve
  await Promise.allSettled(promisesArray);
};

// create the python process
const createPyProc = async () => {
  let script = getScriptPath();

  await killAllPreviousProcesses();

  fp(startingPort, startingPort + portRange)
    .then(([freeport]) => {
      console.log(`Found a free port at ${freeport}. Using port ${freeport} for python process.`);
      pyPort = freeport;
      global.PY_PORT = pyPort;

      console.log(`Starting python process at ${script}`);
      console.log(`API documentation hosted at http://127.0.0.1:${pyPort}/docs`);
      if (guessPackaged()) {
        pyProc = require("child_process").execFile(script, [pyPort], {
          stdio: "ignore",
        });
      } else {
        pyProc = require("child_process").spawn("python", [script, pyPort], {
          stdio: "ignore",
        });
      }

      if (pyProc != null) {
        console.log("child process success on port " + pyPort);
      } else {
        console.error("child process failed to start on port" + pyPort);
      }
    })
    .catch((err) => {
      console.error(err);
    });
};

// Set up a custom protocol handler for the fairshare:// protocol.
if (process.defaultApp) {
  if (process.argv.length >= 2) {
    app.setAsDefaultProtocolClient("fairshare", process.execPath, [path.resolve(process.argv[1])]);
  }
} else {
  app.setAsDefaultProtocolClient("fairshare");
}

// We should prevent the user from running multiple instances of the app.
const gotTheLock = app.requestSingleInstanceLock();

if (!gotTheLock) {
  app.quit();
} else {
  app.on("second-instance", (_event, args, _workingDirectory) => {
    if (process.platform !== "darwin") {
      // Find the arg that is our custom protocol url and store it
      const deepLinkedUrl = args.find((arg) => arg.startsWith("fairshare://"));

      getAuthUrl(deepLinkedUrl);
    }

    // Someone tried to run a second instance, we should focus our window.
    if (mainWindow) {
      if (mainWindow.isMinimized()) mainWindow.restore();
      mainWindow.focus();
    }
  });
}

// Set the default menu of the application
const menu = Menu.buildFromTemplate(menuTemplate);
Menu.setApplicationMenu(menu);

async function createWindow() {
  // Create the browser window.
  console.log(path.join(__dirname, "Icon.png"));

  mainWindow = new BrowserWindow({
    width: 1200,
    height: 850,
    minWidth: 1200,
    minHeight: 850,
    icon: path.join(__dirname, "Icon.png"),
    show: false,
    webPreferences: {
      // Use pluginOptions.nodeIntegration, leave this alone
      // See nklayman.github.io/vue-cli-plugin-electron-builder/guide/security.html#node-integration for more info
      nodeIntegration: process.env.ELECTRON_NODE_INTEGRATION,
      contextIsolation: !process.env.ELECTRON_NODE_INTEGRATION,
      // nodeIntegration: false,
      // contextIsolation: true,
      enableRemoteModule: true,
      preload: path.join(__dirname, "preload.js"),
    },
  });

  // splash screen

  const splash = new BrowserWindow({
    width: 500,
    height: 500,
    frame: false,
    show: true,
    icon: path.join(__dirname, "Icon.png"),
    alwaysOnTop: true,
    transparent: true,
    hasShadow: false,
    resizable: false,
    movable: false,
  });
  splash.loadURL(path.join("file://", __dirname, "/splash-screen.html"));

  // splash screen end

  mainWindow.once("ready-to-show", () => {
    setTimeout(function () {
      splash.close();
      mainWindow.show();

      // Open the DevTools automatically if developing
      if (process.env.WEBPACK_DEV_SERVER_URL) {
        console.log("Opening dev tools in ready-to-show event...");
        mainWindow.webContents.openDevTools();
      }
    }, 500);
  });

  mainWindow.webContents.setWindowOpenHandler(({ url }) => {
    console.log("Intercepting new browser tab/window", url);

    return {
      action: "allow",
      overrideBrowserWindowOptions: {
        frame: true,
        fullScreenable: true,
        backgroundColor: "white",
        webPreferences: {
          allowRunningInsecureContent: false,
          contextIsolation: true,
          devTools: false,
          disableDialogs: true,
          enableRemoteModule: false,
          experimentalFeatures: false,
          nodeIntegration: false,
          nodeIntegrationInWorker: false,
          nodeIntegrationInSubFrames: false,
          plugins: false,
          sandbox: true,
          webSecurity: true,
        },
      },
    };
  });

  enableWebContents(mainWindow.webContents);

  if (process.env.WEBPACK_DEV_SERVER_URL) {
    // Load the url of the dev server if in development mode
    await mainWindow.loadURL(process.env.WEBPACK_DEV_SERVER_URL);

    if (!process.env.IS_TEST) {
      console.log("Opening dev tools in window handler...");

      mainWindow.webContents.openDevTools({ mode: "detach" });
    }
  } else {
    console.log(`Registering the 'app' protocol...`);
    createProtocol("app");

    // Load the index.html when not in development
    mainWindow.loadURL("app://./index.html");
  }
}

// Close the webserver process on app exit
const exitPyProc = async (main_pid) => {
  console.log("killing python process...");

  await killAllPreviousProcesses();

  if (process.platform == "darwin" || process.platform == "linux") {
    pyProc.kill();
    return new Promise(function (resolve) {
      resolve();
    });
  } else {
    const python_script_name = path.basename(getScriptPath());
    let cleanup_completed = false;
    const psTree = require("ps-tree");
    psTree(main_pid, function (_err, children) {
      let python_pids = children
        .filter(function (el) {
          return el.COMMAND == python_script_name;
        })
        .map(function (p) {
          return p.PID;
        });
      // kill all the spawned python processes
      python_pids.forEach(function (pid) {
        process.kill(pid);
      });
      pyProc = null;
      cleanup_completed = true;
    });
    return new Promise(function (resolve) {
      (function waitForSubProcessCleanup() {
        if (cleanup_completed) return resolve();
        setTimeout(waitForSubProcessCleanup, 30);
      })();
    });
  }
};

// Quit when all windows are closed.
app.on("window-all-closed", () => {
  // clean up - remove the custom protocol handler
  app.removeAsDefaultProtocolClient("fairshare");

  // This might need to be moved into the `will-quit` event. Would need some checking on app updates.
  exitPyProc(process.pid).then(() => {
    app.quit();
  });

  // On macOS it is common for applications and their menu bar
  // to stay active until the user quits explicitly with Cmd + Q
  // if (process.platform !== "darwin") {
  //   app.quit();
  // }
});

app.on("activate", () => {
  // On macOS it's common to re-create a window in the app when the
  // dock icon is clicked and there are no other windows open.
  if (BrowserWindow.getAllWindows().length === 0) createWindow();
});

// This will catch clicks on links such as <a href="fairshare://abc=1">open in foobar</a>
// Protocol handler for macOS
app.on("open-url", (event, deepLinkedUrl) => {
  event.preventDefault();
  console.log("deepLinkingUrl", deepLinkedUrl);

  getAuthUrl(deepLinkedUrl);
});

ipcMain.on("fairshare-auth", async (_event, app) => {
  // adding a session ID to the auth url to prevent unauthorized requests
  const authUrl = `https://auth.fairshareapp.io/${app}?session=${SESSION_ID}`;

  shell.openExternal(authUrl);
});

const getAuthUrl = async (deepLinkedUrl) => {
  let url;
  try {
    url = new URL(deepLinkedUrl);
  } catch (e) {
    return;
  }

  const query = url.searchParams;

  const session_id = query.get("session");

  if (!session_id || (session_id && session_id !== SESSION_ID)) {
    dialog.showErrorBox(
      "Unauthorized request",
      "This request did not originate from FAIRshare. An unauthorized app tried to open FAIRshare from your browser. Please contact the FAIRshare team if you think this is a mistake."
    );

    return;
  }

  if (url && url.protocol === "fairshare:") {
    if (url.hostname === "auth-github") {
      const token = query.get("token");

      mainWindow.webContents.send("github-auth-token", token);
    }

    if (url.hostname === "auth-zenodo") {
      const token = query.get("token");
      const refresh_token = query.get("refresh_token");

      mainWindow.webContents.send("zenodo-auth-token", token, refresh_token);
    }

    if (url.hostname === "auth-figshare") {
      const token = query.get("token");
      const refresh_token = query.get("refresh_token");

      mainWindow.webContents.send("figshare-auth-token", token, refresh_token);
    }
  }
};

// This method will be called when Electron has finished
// initialization and is ready to create browser windows.
// Some APIs can only be used after this event occurs.
app.on("ready", async () => {
  if (isDevelopment && !process.env.IS_TEST) {
    // Install Vue Devtools
    try {
      await installExtension(VUEJS3_DEVTOOLS);
    } catch (e) {
      console.error("Vue Devtools failed to install:", e.toString());
    }
  }

  // Check if the current app has the lock instance
  if (gotTheLock) {
    createPyProc();
    createWindow();
  }
});

autoUpdater.on("update-available", () => {
  mainWindow.webContents.send("update-available", true);
  log.info("update_available");
});

autoUpdater.on("update-downloaded", () => {
  log.info("update_downloaded");
  mainWindow.webContents.send("update-downloaded", true);
});

ipcMain.on("check-for-updates", async (_event, channel = "") => {
  // Check for app updates
  if (channel !== "") {
    autoUpdater.channel = channel;
  }
  autoUpdater.checkForUpdatesAndNotify();
});

ipcMain.on("restart-fairshare-for-update", (_event) => {
  exitPyProc(process.pid).then(() => {
    const isSilent = true;
    const isForceRunAfter = true;
    autoUpdater.quitAndInstall(isSilent, isForceRunAfter);
  });
});

// Exit cleanly on request from parent process in development mode.
if (isDevelopment) {
  if (process.platform === "win32") {
    process.on("message", (data) => {
      if (data === "graceful-exit") {
        app.quit();
      }
    });
  } else {
    process.on("SIGTERM", () => {
      app.quit();
    });
  }
}

function retrieveCode(url) {
  return new Promise(function (resolve, reject) {
    let authWindow = new BrowserWindow({
      width: 800,
      height: 600,
      show: false,
      "node-integration": false,
      "web-security": false,
    });

    authWindow.loadURL(url, { userAgent: "Chrome" });
    authWindow.show();
    authWindow.on("closed", () => {
      reject(new Error("closed"));
    });

    function newLink(url) {
      let parsedURL = nodeUrl.parse(url, true);
      let query = parsedURL.query;
      let code = query.code;
      let error = query.error;

      if (error) {
        reject(error);
        authWindow.removeAllListeners("closed");

        setImmediate(function () {
          authWindow.close();
        });
      } else if (code) {
        resolve(code);
        authWindow.removeAllListeners("closed");

        setImmediate(function () {
          authWindow.close();
        });
      }
    }

    // probably needs to add an additional listener to make zenodo oauth working
    // see here https://www.electronjs.org/docs/latest/api/web-contents
    authWindow.webContents.on("will-navigate", (_event, url) => {
      console.log("will-navigate", url);
      newLink(url);
    });

    authWindow.webContents.on("did-get-redirect-request", (_event, _oldUrl, newUrl) => {
      console.log("did-get-redirect-request");
      newLink(newUrl);
    });
  });
}

ipcMain.on("OAuth-Zenodo", async (_event, _test) => {
  // complete this part to make zenodo oauth working
  // api calls are given at the bottom of your zenodo app page
  let success = false;
  const CLIENT_ID_ZENODO = "";
  const CLIENT_SECRET_ZENODO = "";

  await axios
    .get("https://sandbox.zenodo.org/oauth/authorize", {
      params: {
        client_id: CLIENT_ID_ZENODO,
        response_type: "code",
        redirect_uri: "http://localhost:8080",
        scope: "deposit:write deposit:actions",
      },
    })
    .then(async (responseCode) => {
      let authUrl = responseCode.request.res.responseUrl;
      await retrieveCode(authUrl).then(async (code) => {
        await axios
          .post(
            "https://sandbox.zenodo.org/oauth/token",
            {
              client_id: CLIENT_ID_ZENODO,
              client_secret: CLIENT_SECRET_ZENODO,
              grant_type: "authorization_code",
              code: code,
            },
            {
              headers: {
                Accept: "application/json",
              },
            }
          )
          .then(async (response) => {
            mainWindow.webContents.send("OAuth-Zenodo-Reply", response.data.access_token);
            success = true;
          })
          .catch(() => {
            console.log("request token error: ");
          });
      });
    })
    .catch(() => {
      console.log("request code error: ");
    });
  if (!success) {
    mainWindow.webContents.send("OAuth-Zenodo-Reply", "failed");
  }

  mainWindow.webContents.session.clearStorageData();
});
